Optimise current implementation of processing newly added job frameworks

Horiodino opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Added watchCRDs: Introduced a function to monitor CRDs using the apiextensions/v1 API, triggering waitForAPI only when a CRD becomes established.
Integrated watchCRDs into SetupControllers: Spawned a goroutine to watch CRDs and feed discovered GroupVersionKind (GVK) into the discoveredCRDs channel.
Updated setupControllersFromDiscoveredCRDs  the logic to dynamically match and register GVKs with integrations based on incoming discoveredCRDs.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #2837

#### Special notes for your reviewer:
Draft PR in progress – but feel free to provide any suggestions!

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

mimowo commented:
@Horiodino great to see you working on this. Watching the CRDs with the controlller-runtime informer is definitely an option.
cc @varshaprasad96 @ChristianZaccaria @mbobrovskyi 
it would be great if you can help with the review.

mimowo commented:
/retitle Optimise waiting for the availability of CRD-based frameworks

mbobrovskyi commented:
/ok-to-test

Horiodino commented:
ahhh im not sure why most of e2e failing ,locally all succeeded. need some assistance
```
Ran 34 of 34 Specs in 23.627 seconds
SUCCESS! -- 34 Passed | 0 Failed | 0 Pending | 0 Skipped


Ginkgo ran 24 suites in 10m48.689090589s
Test Suite Passed
/home/yoi/kueue/bin/ginkgo-top -i /home/yoi/kueue/bin/integration.json > /home/yoi/kueue/bin/integration-top.yaml
```

mbobrovskyi commented:
> ahhh im not sure why most of e2e failing ,locally all succeeded. need some assistance
> 
> ```
> Ran 34 of 34 Specs in 23.627 seconds
> SUCCESS! -- 34 Passed | 0 Failed | 0 Pending | 0 Skipped
> 
> 
> Ginkgo ran 24 suites in 10m48.689090589s
> Test Suite Passed
> /home/yoi/kueue/bin/ginkgo-top -i /home/yoi/kueue/bin/integration.json > /home/yoi/kueue/bin/integration-top.yaml
> ```

Try to rebase main branch.

Horiodino commented:
ok ,  updated 
```
✓  pkg/util/wait (1.147s) (coverage: 0.1% of statements in ./...)
✓  pkg/webhooks (1.165s) (coverage: 2.0% of statements in ./...)
✓  pkg/workload (1.288s) (coverage: 2.8% of statements in ./...)
✓  pkg/visibility/api/v1beta1 (1.33s) (coverage: 2.0% of statements in ./...)
✓  pkg/scheduler (8.356s) (coverage: 14.1% of statements in ./...)

DONE 2220 tests in 27.119s
```

```
Ran 33 of 33 Specs in 23.729 seconds
SUCCESS! -- 33 Passed | 0 Failed | 0 Pending | 0 Skipped


Ginkgo ran 22 suites in 11m55.637279503s
Test Suite Passed
/home/yoi/kueue/bin/ginkgo-top -i /home/yoi/kueue/bin/integration.json > /home/yoi/kueue/bin/integration-top.yaml
```

ChristianZaccaria commented on pkg/controller/jobframework/setup.go:
```diff
@@ -51,57 +51,71 @@ var (
+
+				var matchedName string
+				var matchedCallbacks *IntegrationCallbacks
+				m.mu.Lock()
+				for name, cb := range m.integrations {
```
This will block access to `m.integrations` by other go routines. Given it's a `read` operation, this shouldn't block other processes.

Horiodino replied:
Thanks for pointing that out , I’ll take care of it.

ChristianZaccaria commented on pkg/controller/jobframework/setup.go:
```diff
@@ -51,57 +51,71 @@ var (
+				for name, cb := range m.integrations {
+					if options.EnabledFrameworks.Has(name) {
+						if name == fmt.Sprintf("%s/%s", currentGVK.Group, strings.ToLower(currentGVK.Kind)) ||
+							strings.EqualFold(reflect.TypeOf(cb.JobType).String(), currentGVK.Kind) {
```
The constant use of `reflect` could cause a slower runtime performance. I'd suggest to use pre-computed strings for direct string comparison.

ChristianZaccaria reviewed (changes requested):
Hi, thanks for your work on this. I left an initial review and some questions.

Horiodino commented on pkg/controller/jobframework/setup.go:
```diff
@@ -51,57 +51,71 @@ var (
+func (m *integrationManager) setupControllersFromDiscoveredCRDs(ctx context.Context, mgr ctrl.Manager, log logr.Logger, discoveredCRDs <-chan schema.GroupVersionKind, opts ...Option) error {
 	options := ProcessOptions(opts...)
 
-	if err := m.checkEnabledListDependencies(options.EnabledFrameworks); err != nil {
```
sorry , i removed that (not intentionally ) forgot to add it again

Horiodino commented on pkg/controller/jobframework/setup.go:
```diff
@@ -51,57 +51,71 @@ var (
 // certs are all in place.
 func SetupControllers(ctx context.Context, mgr ctrl.Manager, log logr.Logger, opts ...Option) error {
-	return manager.setupControllers(ctx, mgr, log, opts...)
+	discoveredCRDs := make(chan schema.GroupVersionKind, 10)
```
already done that not commited yet.

Horiodino commented on pkg/controller/jobframework/setup.go:
```diff
@@ -51,57 +51,71 @@ var (
-	}
-
-	for fwkName := range options.EnabledExternalFrameworks {
-		if err := RegisterExternalJobType(fwkName); err != nil {
```
same : i removed that (not intentionally ) forgot to add it again

Horiodino commented on pkg/controller/jobframework/setup.go:
```diff
@@ -174,3 +188,59 @@ func SetupIndexes(ctx context.Context, indexer client.FieldIndexer, opts ...Opti
+			for _, condition := range crd.Status.Conditions {
+				if condition.Type == apiextensionsv1.Established &&
+					condition.Status == apiextensionsv1.ConditionTrue {
+					go waitForAPI(ctx, mgr, log, gvk, func() {
```
I didn’t think about that initially. This only triggers created or updated were considered, so I didn’t think it could be an issue. But now I believe it might be — thanks for pointing it out.

`We can specifically target enabledIntegrations GVKs to limit the number of concurrent goroutines`. Is options.EnabledExternalFrameworks the correct reference for this?

Another approach we can take is to add a pooling mechanism to better control the goroutines.

ChristianZaccaria replied:
I believe `options.EnabledFrameworks` would be the right reference. Here's the [list of possible frameworks that can be enabled](https://github.com/kubernetes-sigs/kueue/blob/main/apis/config/v1beta1/configuration_types.go#L320-L336). All the frameworks listed there will have their own controller and a webhook in Kueue, which should start only after the CRDs are present.

Pooling sounds like a good approach, thanks!

Horiodino replied:
gotcha  :+1:

Horiodino commented:
```

  Timeline >>
  STEP: tearing down the test environment @ 02/18/25 11:32:27.825
  << Timeline
------------------------------
[AfterSuite] PASSED [1.132 seconds]
[AfterSuite] 
/home/yoi/kueue/test/integration/singlecluster/webhook/jobs/suite_test.go:69

  Timeline >>
  STEP: tearing down the test environment @ 02/18/25 11:32:27.926
  << Timeline
------------------------------
[ReportAfterSuite] PASSED [1.352 seconds]
[ReportAfterSuite] Autogenerated ReportAfterSuite for --json-report --junit-report
autogenerated by Ginkgo
------------------------------

Ran 33 of 33 Specs in 21.394 seconds
SUCCESS! -- 33 Passed | 0 Failed | 0 Pending | 0 Skipped


Ginkgo ran 22 suites in 9m47.046742807s
Test Suite Passed
/home/yoi/kueue/bin/ginkgo-top -i /home/yoi/kueue/bin/integration.json > /home/yoi/kueue/bin/integration-top.yaml
```

Horiodino commented:
can  anyone test it locally im not sure , everything seems to be fine
```

Ran 33 of 33 Specs in 21.394 seconds
SUCCESS! -- 33 Passed | 0 Failed | 0 Pending | 0 Skipped


Ginkgo ran 22 suites in 9m47.046742807s
Test Suite Passed
/home/yoi/kueue/bin/ginkgo-top -i /home/yoi/kueue/bin/integration.json > /home/yoi/kueue/bin/integration-top.yaml

```

PBundyra commented:
Hi @Horiodino, I'm ready to review this PR but I'd like to ensure test are passing first. Please let me know when the PR is ready

Horiodino commented:
is there any guide how i can run these failing test locally?

PBundyra commented:
> is there any guide how i can run these failing test locally?

While there is no a guide for that, if you want to run e2e tests locally you can use `make test-e2e` command from `Makefile`.  There's also an env variable `E2E_TARGET` that allows you to specify a single test scenario. Let me know if you need further guidance

Horiodino commented:
done already , that's the concern , its passing all e2e locally but not in pr-tests.

PBundyra commented:
What command do you use to run those test locally? Does the setup differ from the one on our CI/CD pipeline?

Horiodino commented:
ran all
```
Tests
  test                      Run tests.
  test-integration          Run integration tests for all singlecluster suites.
  test-multikueue-integration  Run integration tests for Multikueue suite.
```

Horiodino commented:
ran make 
```
  test                      Run tests.
  test-integration          Run integration tests for all singlecluster suites.
  test-multikueue-integration  Run integration tests for Multikueue suite.
```

Horiodino commented:
ok , better than last time , don't review it now , i need to add some more fixes, and fix some failing tests.

Horiodino commented:
@PBundyra can you create a issue for adding some docks for the e2e test? they are not in make help option.

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -51,10 +56,22 @@ var (
 
-func (m *integrationManager) setupControllers(ctx context.Context, mgr ctrl.Manager, log logr.Logger, opts ...Option) error {
+func (m *integrationManager) setupControllersFromDiscoveredCRDs(ctx context.Context, mgr ctrl.Manager, log logr.Logger, discoveredCRDs chan schema.GroupVersionKind, opts ...Option) error {
 	options := ProcessOptions(opts...)
```
Do we need this if we process options in `SetupControllers` function?

Horiodino replied:
It's wrong, actually — I'm fixing things up and will answer all questions once I fix up.

Horiodino replied:
It's wrong, actually — I'm fixing things up and will answer all questions once I fix up.

PBundyra replied:
Since we added it here: https://github.com/kubernetes-sigs/kueue/pull/4059/files#diff-9b0c3819f5a2c007ac714ec151937ce08cc2b768527cc28d20580a57f47ca5a7R61
I wondering if it's a duplicate

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -174,3 +298,104 @@ func SetupIndexes(ctx context.Context, indexer client.FieldIndexer, opts ...Opti
+	crdClient, err := clientset.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		log.Error(err, "Failed to create CRD client")
+		close(discoveredCRDs)
```
Can we maybe use `defer` for closing channels to avoid repetitions?

PBundyra commented on pkg/controller/jobframework/setup.go:
Can we add some level of verbosity to logs (`log.V(n).Info`) instead of using the default one?

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -174,3 +298,104 @@ func SetupIndexes(ctx context.Context, indexer client.FieldIndexer, opts ...Opti
+}
+
+func (m *integrationManager) setupStaticControllers(mgr ctrl.Manager, log logr.Logger, opts ...Option) error {
+	options := ProcessOptions(opts...)
```
Do we need to process it once again?

Horiodino replied:
It registers controllers and webhooks for these static types, ensuring they’re ready to manage resources (e.g., Jobs) immediately upon system startup.

See unlike dynamic types (CRDs), which can be discovered later, static types require instant availability. Without this setup, the system can’t process them.

See if we remove setupStaticControllers, controllers for static types aren’t registered. For example:
An E2E test creating a Job expects a Workload resource to be generated.
Without the controller, the Job isn’t reconciled, no Workload is created, and the test fails (Resource Not Found Err)


The separation of static and dynamic type handling is intentional, The static  is handled by setupStaticControllers for immediate .Dynamic Types: Managed via CRD discovery at runtime.
static types don’t need CRD watching, so setupStaticControllers is essential.

Horiodino replied:
Does is make any sense ?

Horiodino replied:
cc @PBundyra —  let me know if this is correct. I’m happy to apply any changes you suggest.

PBundyra replied:
Sorry, I must have missed that. I didn't mean to remove the whole function, but just processing options

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -174,3 +298,104 @@ func SetupIndexes(ctx context.Context, indexer client.FieldIndexer, opts ...Opti
+			return fmt.Errorf("%s: failed to get GVK: %w", fwkNamePrefix, err)
+		}
+		log.Info("Processing static integration", "name", name, "gvk", gvk)
+		if !strings.HasSuffix(gvk.Group, ".k8s.io") || gvk.Group == "batch" {
```
Could you explain what is this check for?

Horiodino replied:
Apologies for the delayed response. I realized I was following the wrong approach earlier. The original watchCRDs might have missed the kueue.x-k8s.io/Workload CRD if the informer failed to sync  possibly due to RBAC or network issues  without providing clear feedback. I’ve tested thoroughly and made a lot of changes. I’ll push the updates ASAP.

Horiodino replied:
It's wrong, actually — I'm fixing things up and will answer all questions once I fix up.

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -66,45 +83,152 @@ func (m *integrationManager) setupControllers(ctx context.Context, mgr ctrl.Mana
+	retries := 0
+	backoff := setupRetryDelay
+
+	for retries < maxSetupRetries {
```
Can we initialize and increment `retries` within the loop statement?

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -66,45 +83,152 @@ func (m *integrationManager) setupControllers(ctx context.Context, mgr ctrl.Mana
+			"attempt", retries,
+			"maxRetries", maxSetupRetries)
+
+		if retries >= maxSetupRetries {
+			log.Error(err, "Maximum controller setup retries reached",
+				"gvk", gvk,
+				"integration", name)
+			return
+		}
```
Isn't this code dead because of the for loop condition?

PBundyra commented on pkg/controller/jobframework/setup.go:
```diff
@@ -66,45 +83,152 @@ func (m *integrationManager) setupControllers(ctx context.Context, mgr ctrl.Mana
 
+// verifyAPIReady checks if the API server is ready to serve resources of the given GVK
+func verifyAPIReady(ctx context.Context, mgr ctrl.Manager, c client.Client, scheme *runtime.Scheme, gvk schema.GroupVersionKind) error {
+	if err := restMappingExists(mgr, gvk); err != nil {
+		return fmt.Errorf("REST mapping not available: %w", err)
+	}
+
```
Do we need this if `waitForAPI` already does that?

PBundyra commented on pkg/controller/jobframework/setup.go:
Can we please switch loggers to depth of at least 2? We usually don't use depth of 0 or 1 in Kueue

Horiodino commented:
Hey @PBundyra, I'm running into an issue and could use your help!  

I have `setupStaticControllers` configured to handle only static resources like Pods, Deployments, and StatefulSets (from core, apps, and batch groups), but not dynamic ones like custom CRDs. The problem is that static resources aren’t working because Workload objects aren’t being created for them (e.g., `statefulset-sts-18b6e` not found), likely due to controller setup failures.  

Previously, my function used a condition:  
```go
!strings.HasSuffix(gvk.Group, ".k8s.io") || gvk.Group == "batch"
```  
which was too broad—it incorrectly processed both static and dynamic types but at least it was working. Now, after refining the condition, I'm facing E2E test failures where Pods, Deployments, and StatefulSets aren't found.  

This is my current implementation:  
```go
func (m *integrationManager) setupStaticControllers(mgr ctrl.Manager, log logr.Logger, opts ...Option) error {
    options := ProcessOptions(opts...)
    builtInGroups := sets.NewString("", "apps", "batch")
    return m.forEach(func(name string, cb IntegrationCallbacks) error {
        if !options.EnabledFrameworks.Has(name) {
            return nil
        }
        fwkNamePrefix := fmt.Sprintf("jobFrameworkName %q", name)
        gvk, err := apiutil.GVKForObject(cb.JobType, mgr.GetScheme())
        if err != nil {
            return fmt.Errorf("%s: failed to get GVK: %w", fwkNamePrefix, err)
        }
        log.Info("Processing integration", "name", name, "gvk", gvk)
        if builtInGroups.Has(gvk.Group) {
            if err := restMappingExists(mgr, gvk); err == nil {
                log.Info("API exists, setting up static controller", "gvk", gvk)
                return m.setupControllerAndWebhook(mgr, name, fwkNamePrefix, cb, options, opts...)
            } else if !meta.IsNoMatchError(err) {
                return fmt.Errorf("%s: failed to verify API: %w", fwkNamePrefix, err)
            }
            log.Info("API not available for static type, skipping", "gvk", gvk)
        } else {
            log.Info("Skipping non-static integration", "name", name, "gvk", gvk)
        }
        return nil
    })
}
```  
Do you see anything off here? Any ideas on why this might be causing missing Pod/Deployment/StatefulSet issues?

Horiodino commented on pkg/controller/jobframework/setup.go:
```diff
@@ -180,3 +308,107 @@ func SetupIndexes(ctx context.Context, indexer client.FieldIndexer, opts ...Opti
+	crdInformer := factory.Apiextensions().V1().CustomResourceDefinitions().Informer()
+
+	_, err = crdInformer.AddEventHandler(tools.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
```
Do we need to explicitly define delete? like here we did , I fixed all failures, but during `leaderworkerset_test`, it's failing at delete. I think that might be the reason.

I'm slightly confused—since it's a workload, shouldn't it be handled by the workload controller?

Horiodino replied:
No problem, I got it. It's the leader-workers test controller.

mbobrovskyi commented:
/release-note-edit
```release-note
NONE
```

tenzen-y reviewed (commented):
I don't think that this is a proper approach.
This tries to re-implement reconcile mechanism
We really should not re-implement existing mechanism by ourselves.

Horiodino commented:
> I don't think that this is a proper approach. This tries to re-implement reconcile mechanism We really should not re-implement existing mechanism by ourselves.

When constructing a workload, we need to specify the owner reference; otherwise, the associated resources won't be deleted properly. Additionally, the Reconciler is not handling the scale-down properly, which could be causing the test failure—that's why I added it. Try running it locally with the Kueue main branch's leader Reconciler along with the current changes, and you'll see some errors.  

If you have a better approach, please suggest.

PBundyra commented on pkg/controller/jobs/leaderworkerset/leaderworkerset_reconciler.go:
Are those changes necessary for the primary issue we're trying to solve here?

Horiodino replied:
Previously, we weren't using a CRD, so this behavior wasn't explicitly handled. Now that we're managing state via a CRD, we need to handle scale-down logic  ourselves to avoid leaving extra Workloads running.
With this change, all test cases pass.
Open to suggestions if there's a cleaner way to approach this.

Horiodino replied:
Hey everyone, Could someone please provide an update or feedback? I’m concerned the issue might go stale if it doesn't get attention soon.

mimowo replied:
Yes, I agree with @PBundyra these changes to LWS are not related to the PR. They may or may not be ok, but we should focus the PR on detecting the CRDs as they get installed. Decreasing the size of the PR will make it much easier to progress with review.

Horiodino replied:
> Yes, I agree with @PBundyra these changes to LWS are not related to the PR. They may or may not be ok, but we should focus the PR on detecting the CRDs as they get installed. Decreasing the size of the PR will make it much easier to progress with review.

I strongly agree with your point. That said, the `LeaderWorkerSet` changes are essential to adapt the controller to the CRD-based approach and to get the tests passing.

Splitting the changes might not be the best path forward, as it results in test failures. It would be better to merge [[PR #4790]first before introducing changes related to the CRD watching mechanism.

We can wait for that PR to stabilize , I’m  seeing just flaky tests there and then move ahead with this issue.

mimowo replied:
> the LeaderWorkerSet changes are essential to adapt the controller to the CRD-based approach and to get the tests passing.

This is surprising to me, maybe I'm missing something. I thought this PR conceptually should delay calling `waitForAPI` until we observe the CRD to be installed (based on the informer on the CRDs). Could you double check if the other changes are really necessary, and if so, summarize why?

PBundyra commented on config/components/rbac/role.yaml:
```diff
@@ -347,3 +347,21 @@ rules:
+- apiGroups: [ "apiextensions.k8s.io" ]
+  resources: [ "customresourcedefinitions" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
+- apiGroups: [ "kueue.x-k8s.io" ]
+  resources: [ "workloads" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
+- apiGroups: [ "" ]
+  resources: [ "pods" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
```
Isn't this is a duplicate to what's already in the file

PBundyra commented on config/components/rbac/role.yaml:
```diff
@@ -347,3 +347,21 @@ rules:
   - get
   - patch
   - update
+- apiGroups: [ "apiextensions.k8s.io" ]
+  resources: [ "customresourcedefinitions" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
```
Why do we need this?

mimowo replied:
I think the "read" rbac might be needed by Kueue to be able to watch the CRD definitions. However, almost certainly we don't need the "write" rbac - create update and delete.

Also, ideally, these should be genereted based on kube-builder marker comments, rather than specified manually, [example](https://github.com/kubernetes-sigs/kueue/blob/e9b35497ccf5b0534cce64a2a5f71c81b0926d6d/pkg/controller/jobs/job/job_controller.go#L72-L81)

PBundyra commented on config/components/rbac/role.yaml:
```diff
@@ -347,3 +347,21 @@ rules:
+- apiGroups: [ "" ]
+  resources: [ "pods" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
+- apiGroups: [ "leaderworkerset.x-k8s.io" ]
+  resources: [ "leaderworkersets" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
+- apiGroups: [ "apps" ]
+  resources: [ "deployments", "statefulsets" ]
+  verbs: [ "get", "list", "watch", "create", "update", "delete" ]
```
Please follow formatting present in the file

tenzen-y commented:
> When constructing a workload, we need to specify the owner reference; otherwise, the associated resources won't be deleted properly.

I could not understand what you wanted to say. Why can't we just watch CRDs?

>  Additionally, the Reconciler is not handling the scale-down properly, which could be causing the test failure—that's why I added it.

We have a leader-election mechanism https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/leader_aware_reconciler.go. So, I can not understand what you wanted to say.

Horiodino commented:
> > When constructing a workload, we need to specify the owner reference; otherwise, the associated resources won't be deleted properly.
> 
> I could not understand what you wanted to say. Why can't we just watch CRDs?
> 
> > Additionally, the Reconciler is not handling the scale-down properly, which could be causing the test failure—that's why I added it.
> 
> We have a leader-election mechanism https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/leader_aware_reconciler.go. So, I can not understand what you wanted to say.

I've removed the changes I made to the LeaderWorkerSet reconcile logic. This should give you more context about what I was referring to earlier. For now, please ignore the RBAC configuration ,I'll fix that once the main issue is resolved. Take a look at the test case for additional context; it should help clarify what I'm trying to convey.
